### DISCUSSION TITLE
Trying to understand AssemblyFixture

### DISCUSSION THREAD

**Discussion by PureKrome** [xunit/xunit#3080] [Question] [answered]
👋🏻 G'Day xunit team.

I'm trying to play with the new v3 library.

I have a number of tests which all hit a DB. As such, I wish to use TestContainers.

I was thinking of using `AssemblyFixture` to create the Database TestContainer instance _once_ (for this entire test-run):

```
[assembly: AssemblyFixture(typeof(SqlServerFixture))]
```

Then each _class_ will create it's own ConnectionString (which means, it has it's own db "tenant" in this single TestContainer instance).

Here's some code to try and explain my poor explanation:

```
[assembly: AssemblyFixture(typeof(SqlServerFixture))]

public class MyTests
{
    private readonly string _connectionString;

    public MyTests(SqlServerFixture sqlServerFixture, ITest test)
    {
        _connectionString = sqlServerFixture.CreateConnectionString(test.TestDisplayName);
    }

    [Fact]
    public async Task SomeTestAsync()
    {
        // Some useless code which hits a db. The db happens to be via a test container.
        using var db = new SqlConnection(_connectionString);
        var result = await db.SingleOrDefaultAsync("SELECT Blah...");
    }
}
```

- DB test container is created once.
- Each method in this class `MyTests` will end up creating a connection string based of the name of the test that is being executed.

Can xunit.v3 do this? 
Do I need to use an IClassFixture?

I do not want to use an ICollectionFixture because that means everything in that _named collection_ will run synchronously. By splitting my tests into classes, they start to run _more parallel_.

NOTE: this is similar (and inspired from) https://github.com/xunit/xunit/discussions/3017 but not actually the same as that.

**Comment by bradwilson** [comment, marked as answer]
Yes, you should be able to do that. The constructor is the correct place to do this so that it's done once for each test (or, if you need to use async methods during initialization, you can implement `IAsyncLifetime`).

```csharp
using Xunit;

[assembly: AssemblyFixture(typeof(SqlServerFixture))]

public class SqlServerFixture
{
    public string CreateConnectionString(string testDisplayName) =>
        $"Connection string for {testDisplayName}";
}

public class UnitTest
{
    string connectionString;

    public UnitTest(SqlServerFixture fixture)
    {
        connectionString = fixture.CreateConnectionString(TestContext.Current.Test!.TestDisplayName);
    }

    [Fact]
    public void TestMethod() =>
        Assert.Equal("Connection string for UnitTest.TestMethod", connectionString);
}
```

Note that TestContext is usable in the entirety of the testing pipeline, so `Test` is marked as nullable for when it's called outside the context of a test. This code will never be called outside the context of a test, so I just use the `!` operator to ignore the null possibility.

**Reply by PureKrome** [reply]
Thank you @bradwilson ! This is lovely. I'll give it a go tonight and test this out. Also appreciate
- Adding a TestContext in v3.
- Explaining the `!` in the above scenario.

❤️

**Reply by PureKrome** [reply]
Yep - this works perfectly! win! now onto other test performance issues (me, not xunit.v3).
